### PR TITLE
Implementation of validator disabling strategy in the runtime

### PR DESCRIPTION
**Pull request by tdimitrov**
This PR is a recreation of https://github.com/paritytech/polkadot-sdk/pull/1214 which after the transfer includes some rogue commits. Fixing it is too much hassle (there are additional merges after that) so I'm recreating it.

The strategy is described in https://github.com/paritytech/polkadot-sdk/issues/784#issuecomment-1732644576. This PR outlines an initial implementation with todos.
Related to https://github.com/paritytech/polkadot-sdk/issues/784
Link to the PR in the old substrate repo: https://github.com/paritytech/substrate/pull/14697

Things that needs to be addressed:

- [ ] Fix the TODOs in the code.
- [ ] Add tests
- [ ] Add runtime migration for staking palet
- [ ] Add benchamarks

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4224719

File: substrate/frame/staking/src/slashing.rs [hunk 1]
```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
...
+			Err(pos) => {
+				offending.insert(pos, (validator_index_u32, slash_proportion));
 
 				let offending_threshold =
 					T::OffendingValidatorsThreshold::get() * validators.len() as u32;
 
 				if offending.len() >= offending_threshold as usize {
-					// force a new era, to select a new validator set
+					// force a new era, to select a new validator set. The era will be forced on the
+					// next session
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:354]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283242773

File: substrate/frame/staking/src/slashing.rs [hunk 2]
```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
...
+// Gets a Vec and its desired length as an input. Randomly removes values from the Vec until its
+// length matches the desired length. The validator length should be bigger than the desired one.
+fn remove_random_validators<T: Config>(validators: Vec<u32>, desired_len: usize) -> Vec<u32> {
+	debug_assert!(validators.len() > desired_len);
+
+	let entropy = compute_entropy::<T>(<frame_system::Pallet<T>>::parent_hash());
+	let mut rng = rand_chacha::ChaChaRng::from_seed(entropy.into());
+
+	let mut indecies_to_remove = BTreeSet::new();
+	for _ in 0..=desired_len.saturating_sub(validators.len()) {
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:420]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283287217

File: substrate/frame/staking/src/slashing.rs [hunk 3]
```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
...
+
+	// Now selectively disable based on the offence
+	let mut disabled = Vec::new();
+	for (v, o) in current_offenders {
+		if disable_rand::<T>(o) {
+			disabled.push(*v);
+		}
+	}
+
+	let disabled = if disabled.len() > LIMIT {
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:464]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283303368

**Pull request closed without merging**